### PR TITLE
Replace manual polling with a Queue.Queue in the backend

### PR DESCRIPTION
**Pull request by caisq**
So there is no need to call time.sleep() in while loops in
debugger_plugin.py anymore, because queue.get() will block if the queue
is empty.

### CODE REVIEW

File: tensorboard/plugins/debugger/debugger_plugin.py [hunk 1]
```diff
@@ -540,56 +539,8 @@ def _serve_run_keys_to_graphs(self, request):
...
-          request,
-          {
-              "state": "tensors",
-              "data": {new_key: tensor_value},
-          },
-          "application/json")
-    else:
-      raise ValueError("Invalid value in _debugging_connection_state: %s" %
-                       self._debugging_connection_state)
+    comm_data = self._debugger_data_server.comm_queue.get()
```
Comment (chihuahua) [inline tensorboard/plugins/debugger/debugger_plugin.py]: Cool! Maybe add a comment noting that this blocks if `comm_queue` is empty?

File: tensorboard/plugins/debugger/debugger_server_lib.py [hunk 2]
```diff
@@ -213,21 +245,16 @@ def on_value_event(self, event):
...
-      tensor_value_str = "%s: %s" % (node_name, tensor_value)
-      tf.logging.warning(
-          "Recording tensor value in _comm_data: %s", tensor_value_str)
-      self._comm_data[event.wall_time] = tensor_value_str
+      self._comm_queue.put(_comm_tensor_data(event))
 
       # Wait for acknowledgement.
       self._ack["client"] = False
+      tf.logging.warning("Waiting for client ack (tensors)...")
       while not self._ack["client"]:
```
Comment (chihuahua) [inline tensorboard/plugins/debugger/debugger_server_lib.py]: So every time the user hits 'continue', we create a new `DebuggerDataStreamHandler`?

File: tensorboard/plugins/debugger/debugger_server_lib.py [hunk 3]
```diff
@@ -40,6 +40,47 @@
...
+          "fetch_names": fetch_names,
+          "target_names": target_names,
+          "run_key": _run_key(feed_names, fetch_names, target_names),
+      }
+  }
+
+
+def _comm_tensor_data(event):
+  tensor_value = tf.make_ndarray(event.summary.value[0].tensor)
+  # TODO(cais): The value of the dict should be tensor keys, instead of
```
Comment (chihuahua) [inline tensorboard/plugins/debugger/debugger_server_lib.py]: OK, and then we'd let the frontend request the tensor via the key via a separate route?

File: tensorboard/plugins/debugger/debugger_server_lib.py [hunk 4]
```diff
@@ -40,6 +40,47 @@
...
+  # the value or a string representation of it, which is too big for
+  # many real-life tensors.
+  node_name, output_slot, debug_op = event.summary.value[0].node_name.split(":")
+  output_slot = int(output_slot)
+  tf.logging.warning(
+      "Recording tensor value: %s, %d, %s", node_name, output_slot, debug_op)
+  tensor_value = tf.make_ndarray(event.summary.value[0].tensor)
+  tensor_value_str = str(tensor_value)
```
Comment (chihuahua) [inline tensorboard/plugins/debugger/debugger_server_lib.py]: So, does this look like the JSON representation? ie, [42.0, 1337.42, 3.14]?

File: tensorboard/plugins/debugger/debugger_server_lib.py [hunk 5]
```diff
@@ -40,6 +40,47 @@
...
+  return {
+      "type": "tensor",
+      "timestamp": event.wall_time,
+      "data": {
+          "node_name": node_name,
+          "output_slot": output_slot,
+          "debug_op": debug_op,
+          "value": tensor_value_str,
```
Comment (chihuahua) [inline tensorboard/plugins/debugger/debugger_server_lib.py]: If we send the type of the tensor (float, int, bool, string, etc), the frontend could probably better handle the value. TensorProto has a dtype field: https://github.com/tensorflow/tensorflow/blob/a41b510f0358d1212d2c548d65c4b1c5e4aa01c0/tensorflow/core/framework/tensor.proto#L15

The `DataType` field is an enum int.
https://github.com/tensorflow/tensorflow/blob/a41b510f0358d1212d2c548d65c4b1c5e4aa01c0/tensorflow/core/framework/types.proto#L10

File: tensorboard/plugins/debugger/debugger_server_lib.py (hunk 3 above)
Comment (caisq) [reply tensorboard/plugins/debugger/debugger_server_lib.py]: Correct,

File: tensorboard/plugins/debugger/debugger_server_lib.py (hunk 4 above)
Comment (caisq) [reply tensorboard/plugins/debugger/debugger_server_lib.py]: Unfortunately no. The `__str__()` of a numpy.ndarray omits the comma in many cases. 

There might be some hacky solution to get it jsonified, see here: https://stackoverflow.com/questions/26646362/numpy-array-is-not-json-serializable

File: tensorboard/plugins/debugger/debugger_server_lib.py (hunk 2 above)
Comment (caisq) [reply tensorboard/plugins/debugger/debugger_server_lib.py]: No. Every time the user clicks "Continue", a non-long-polling HTTP request is sent to the `/client_ack` route of the TDP backend. There is no new `DebuggerDataStreamHandler` created.

File: tensorboard/plugins/debugger/debugger_server_lib.py (hunk 5 above)
Comment (caisq) [reply tensorboard/plugins/debugger/debugger_server_lib.py]: I added a TODO item.

**Pull request merged**